### PR TITLE
Add support for timeout for restic backup and restore

### PR DESCRIPTION
**Pull request by ishtiaqhimel**

### CODE REVIEW

File: pkg/restic/commands.go [hunk 1]
```diff
@@ -200,7 +209,15 @@ func (w *ResticWrapper) backupFromStdin(options BackupOptions) ([]byte, error) {
 	args = w.appendInsecureTLSFlag(args)
 	args = w.appendMaxConnectionsFlag(args)
 
-	commands = append(commands, Command{Name: ResticCMD, Args: args})
+	command := Command{Name: ResticCMD, Args: args}
+	if w.config.Timeout != nil {
```
Comment (anisurrahman75) [inline pkg/restic/commands.go:213]: We can use a separate method to get modified backup command with timeout. As a result, code reputation will be reduced.

```go
command := Command{Name: ResticCMD, Args: args}
	if w.config.Timeout != nil {
		command = w.getCommandWithTimeOut(args)
	}
```

```go
func (w *ResticWrapper) getCommandWithTimeOut(args []interface{}) Command {
	timeoutArgs := []interface{}{fmt.Sprintf("%f", w.config.Timeout.Seconds())}
	timeoutArgs = append(timeoutArgs, ResticCMD, args)
	timeoutArgs = append(timeoutArgs, args...)
	return Command{Name: TimeoutCMD, Args: timeoutArgs}
}

```

File: pkg/restic/restic_test.go [hunk 2]
```diff
@@ -464,6 +463,38 @@ func TestBackupRestoreWithArgs(t *testing.T) {
 	}
 }
 
+func TestBackupWithTimeout(t *testing.T) {
+	tempDir, err := os.MkdirTemp("", "stash-unit-test-")
```
Comment (anisurrahman75) [inline pkg/restic/restic_test.go:467]:
```go
tempDir, err := os.MkdirTemp("", "kubestash-unit-test-")
```

**Review by anisurrahman75 — CHANGES_REQUESTED** [review]

File: pkg/restic/restic_test.go (hunk 2 above)
Comment (ishtiaqhimel) [reply pkg/restic/restic_test.go:467]: All other tests contain `stash-unit-test-`. Kept this for the `homogeneous` dir name for all tests.

File: apis/core/v1alpha1/backupsession_types.go [hunk 3]
```diff
@@ -75,10 +80,10 @@ type BackupSessionStatus struct {
 	// +optional
 	Duration string `json:"duration,omitempty"`
 
-	// Deadline specifies the deadline of backup. BackupSession will be
-	// considered Failed if backup does not complete within this deadline
+	// Deadline specifies the deadline of backup. Backup will be
+	// considered Failed if it does not complete within this deadline
 	// +optional
-	Deadline *metav1.Time `json:"sessionDeadline,omitempty"`
+	Deadline *metav1.Time `json:"backupDeadline,omitempty"`
```
Comment (hmsayem) [inline apis/core/v1alpha1/backupsession_types.go:86]: fix the json tag.

**Review by hmsayem — CHANGES_REQUESTED** [review]

File: apis/core/v1alpha1/restoresession_types.go [hunk 4]
```diff
@@ -278,10 +278,10 @@ type RestoreSessionStatus struct {
 	// +optional
 	Duration string `json:"duration,omitempty"`
 
-	// Deadline specifies a timestamp till this session is valid. If the session does not complete within this deadline,
-	// it will be considered as failed.
+	// Deadline specifies the deadline of restore. Restore will be
+	// considered Failed if it does not complete within this deadline
 	// +optional
-	Deadline *metav1.Time `json:"deadline,omitempty"`
+	Deadline *metav1.Time `json:"restoreDeadline,omitempty"`
```
Comment (hmsayem) [inline apis/core/v1alpha1/restoresession_types.go:284]: fix the json tag

File: apis/core/v1alpha1/backupsession_types.go (hunk 3 above)
Comment (ishtiaqhimel) [reply apis/core/v1alpha1/backupsession_types.go:86]: Should it be just deadline? If yes, wouldn't it mean the deadline of backupsession?

**Review by hmsayem — APPROVED** [review]

**Pull request merged**